util/diff_runs.py

```diff
@@ -0,0 +1,244 @@
+            if runAfter is None:
+                self.logger.warning('Failed to fetch %s' % specAfter)
+            if runBefore is None or runAfter is None:
+                continue
+
+            differences = 0
+            checked = 0
+            added = 0
+            removed = 0
+
+            for k in filter(lambda x: x not in runAfter, runBefore.keys()):
+                removed += 1
+
+            for test in runAfter.keys():
+                checked += 1
+
+                if test not in runBefore:
+                    added += 1
+                    continue
+
+                resultAfter = runAfter[test]
+                resultBefore = runBefore[test]
+
+                passingDelta = resultAfter[0] - resultBefore[0]
+                totalDelta = resultAfter[1] - resultBefore[1]
+
+                if totalDelta > 0:
+                    self.logger.info('%s has %d new tests (total)'
+                                     % (test, totalDelta))
+                elif totalDelta < 0:
+                    self.logger.info('%s has %d removed tests (total)')
+
+                if passingDelta < 0:
+                    self.logger.warning('%s has %d new failures'
+                                        % (test, math.fabs(passingDelta)))
+                elif passingDelta > 0:
+                    self.logger.info('%s has %d new passes'
+                                     % (test, passingDelta))
+
+                delta = math.fabs(totalDelta) + math.fabs(passingDelta)
+                logging.debug('%s has %d differences (%s vs %s)'
+                              % (test,
+                                 passingDelta,
+                                 resultAfter,
+                                 resultBefore))
+                differences += delta
+
+            self.logger.info(
+                'Finished diff of %s and %s with %d differences in %d tests'
+                % (specBefore, specAfter, differences, checked))
+            if (added > 0):
+                self.logger.info('%d tests ran in %s but not in %s'
+                                 % (added, specAfter, specBefore))
+            if (removed > 0):
+                self.logger.info('%d tests ran in %s but not in %s'
+                                 % (removed, specBefore, specAfter))
+
+
+class Fetcher(object):
+    '''Placeholder for stubbing 'fetchResults' for the unit tests.'''
+
+    def __init__(self):
+        self.pool = urllib3.PoolManager(
+            cert_reqs='CERT_REQUIRED',
+            ca_certs=certifi.where())
+
+    def fetchResults(self, sha, platform):
+        '''Fetch a python object representing the test run results JSON for the
+        given sha/platform spec. '''
+        # type: (str, str) -> object
+
+        encodedArgs = urlencode({'sha': sha, 'platform': platform})
+        url = 'https://wpt.fyi/json?' + encodedArgs
+
+        try:
+            response = self.pool.request('GET', url, redirect=False)
```

Comment:
If this is synchronous, what are we getting from the pooling?

Reply:
cert checks.